src/ostrowski/rationals/bounded.lean

```diff
@@ -0,0 +1,364 @@
+        : by { congr' 2, exact real.rpow_neg (le_of_lt zero_lt_p) 1, }
+      ... = real.log ((p⁻¹) ^ α)
+        : by { congr' 3, exact real.rpow_one p, }
+      ... = real.log ((1/p) ^ α)  : by simp,
+  },
+  {
+    intro h,
+    rw real.one_rpow α,
+
+    -- We reason by contradiction. Our hypothesis becomes `abv q < 1`.
+    by_contradiction h',
+    have abvq_lt_one: abv q < 1,
+    from lt_iff_le_and_ne.2 ⟨ (all_nat_le_one q), h' ⟩,
+    -- We find `N` such that `(abv p) ^ N, (abv q) ^ N < 1/2`.
+    obtain ⟨ n, abvpn_lt_half, abvqn_lt_half ⟩: ∃ n: ℕ, ((abv p)^n < 1/2) ∧ ((abv q)^n < 1/2),
+    {
+      have geom_tendsto_zero: ∀ r: ℝ, (0 ≤ r) → (r < 1) →
+        ∃ N: ℕ, ∀ n ≥ N, r^n < 1/2,
+      {
+        -- Again, some formal play.
+        -- Went actually pretty nicely !
+        intros r h₁ h₂,
+        have p := metric.tendsto_at_top.1 (tendsto_pow_at_top_nhds_0_of_lt_1 h₁ h₂),
+        unfold dist at p,
+        unfold abs at p,
+        specialize p (1/2) (by linarith),
+        rcases p with ⟨ N, hN ⟩,
+        use N,
+        intros n hn,
+        specialize hN n hn,
+        calc r^n = r^n - 0                 : by simp
+          ... ≤ max (r^n - 0) (-(r^n - 0)) : le_max_left _ _
+          ... < 1/2 : hN,
+      },
+      obtain ⟨ N₁, pN₁ ⟩: ∃ N: ℕ, ∀ n ≥ N, (abv p)^n < 1/2,
+      { exact geom_tendsto_zero (abv p) (abv_nonneg abv p) abvp_lt_one },
+      obtain ⟨ N₂, pN₂ ⟩: ∃ N: ℕ, ∀ n ≥ N, (abv q)^n < 1/2,
+      { exact geom_tendsto_zero (abv q) (abv_nonneg abv q) abvq_lt_one },
+      use max N₁ N₂,
+      exact ⟨
+        pN₁ (max N₁ N₂) (by linarith only [le_max_left N₁ N₂]),
+        pN₂ (max N₁ N₂) (by linarith only [le_max_right N₁ N₂])
+      ⟩,
+    },
+    -- We use Bézout to find `u`, `v` such that `1 = u * p^n + v * q^n`.
+    obtain ⟨ u, v, bezout ⟩: ∃ u: ℤ, ∃ v: ℤ, 1 = u * p^n + v * q^n,
+    {
+      have p₁: ((p^n).gcd (q^n): ℤ) = (1: ℤ),
+      {
+        norm_cast,
+        suffices coprimes: p.coprime q,
+        { exact (nat.coprime.pow_left n ∘ nat.coprime.pow_right n) coprimes },
+        rw nat.coprime_primes
+          (nat.prime_iff_prime.2 p_prime)
+          (nat.prime_iff_prime.2 q_prime),
+        exact h,
+      },
+      use [(p^n).gcd_a (q^n), (p^n).gcd_b (q^n)],
+      rw ← p₁,
+      rw mul_comm _ (p^n: ℤ),
+      rw mul_comm _ (q^n: ℤ),
+      norm_cast,
+      exact nat.gcd_eq_gcd_ab (p^n) (q^n),
+    },
+    have abv_rel_le_one: ∀ k: ℤ, abv k ≤ 1,
+    {
+      intro k,
+      by_cases h: 0 < k,
+      {
+        rw ← abs_of_pos h,
+        rw int.abs_eq_nat_abs,
+        exact all_nat_le_one _,
+      },
+      {
+        simp at h,
+        rw ← abv_neg abv,
+        norm_cast,
+        rw ← abs_of_nonpos h,
+        rw int.abs_eq_nat_abs,
+        exact all_nat_le_one _,
+      }
+    },
+    have abv_bezout_lt_one: abv (u * p^n + v * q^n) < 1,
+    {
+      -- Is there a way to simplify this ?
+      -- (Of course these two properties could be gathered in one function...)
```

Comment:
i guess linarith should be able to do it
but i am not really friend with linarith yet

Reply:
I tried `linarith` and `nlinarith` without success.